pkg/runner/test_runner.go

```diff
@@ -6,282 +6,796 @@
-	go func() {
-		log.Infof("running pipeline")
-		errCh <- testPipeline.run(completedTargets)
 	}()
-
-	defer terminateInjection()
-	// Receive targets from the completed channel controlled by the pipeline, while
-	// waiting for termination signals or fatal errors encountered while running
-	// the pipeline.
-	for {
-		select {
-		case err := <-errCh:
-			log.Debugf("test runner terminated, returning %v", err)
-			return err
-		case t := <-completedTargets:
-			log.Infof("test runner completed target: %v", t)
-		}
+	chans := test.TestStepChannels{In: ss.inCh, Out: ss.outCh, Err: ss.errCh}
+	runErr := ss.sb.TestStep.Run(ctx, chans, ss.sb.Parameters, ss.ev)
+	tr.mu.Lock()
+	ss.stepRunning = false
+	if runErr != nil {
+		ss.runErr = runErr
 	}
+	tr.mu.Unlock()
+	// Signal to the result processor that no more will be coming.
+	tr.safeCloseOutCh(ss)
+	ss.log.Debugf("%s: step runner finished", ss)
 }
 
-// NewTestRunner initializes and returns a new TestRunner object. This test
-// runner will use default timeout values
-func NewTestRunner() TestRunner {
-	return TestRunner{
-		timeouts: TestRunnerTimeouts{
-			StepInjectTimeout:   config.StepInjectTimeout,
-			MessageTimeout:      config.TestRunnerMsgTimeout,
-			ShutdownTimeout:     config.TestRunnerShutdownTimeout,
-			StepShutdownTimeout: config.TestRunnerStepShutdownTimeout,
-		},
+// reportTargetResult reports result of executing a step to the appropriate target runner.
+func (tr *TestRunner) reportTargetResult(ctx statectx.Context, ss *stepState, tgt *target.Target, res error) error {
+	resCh, err := func() (chan error, error) {
+		tr.mu.Lock()
+		defer tr.mu.Unlock()
+		ts := tr.targets[tgt.ID]
+		if ts == nil {
+			return nil, fmt.Errorf("%s: result for nonexistent target %s %v", ss, tgt, res)
+		}
+		if ss.tgtDone[tgt] {
+			return nil, &cerrors.ErrTestStepReturnedDuplicateResult{
+				StepName: ss.sb.TestStepLabel,
+				Target:   tgt.ID,
+			}
+		}
+		ss.tgtDone[tgt] = true
+		// Begin is also allowed here because it may happen that we get a result before target runner updates phase.
+		if ts.CurStep != ss.stepIndex || (ts.CurPhase != targetStepPhaseBegin && ts.CurPhase != targetStepPhaseRun) {
+			return nil, &cerrors.ErrTestStepReturnedUnexpectedResult{
+				StepName: ss.sb.TestStepLabel,
+				Target:   tgt.ID,
+			}
+		}
+		if ts.resCh == nil {
+			// This should not happen, must be an internal error.
+			return nil, fmt.Errorf("%s: target runner %s is not there, dropping result on the floor", ss, ts)
+		}
+		ss.log.Debugf("%s: result for %s: %v", ss, ts, res)
+		return ts.resCh, nil
+	}()
+	if err != nil {
+		return err
+	}
+	select {
+	case resCh <- res:
+		break
+	case <-ctx.Done():
+		break
 	}
+	return nil
 }
 
-// NewTestRunnerWithTimeouts initializes and returns a new TestRunner object with
-// custom timeouts
-func NewTestRunnerWithTimeouts(timeouts TestRunnerTimeouts) TestRunner {
-	return TestRunner{timeouts: timeouts}
+func (tr *TestRunner) safeCloseOutCh(ss *stepState) {
```

Comment:
Looks like `safeCloseOutCh` and `safeCloseErrCh` can be merged together and accept an argument from outside.

Reply:
yes, except channels have different types and i couldn't find a way to cast them to something common.